net/ipfamily.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import (
+	"fmt"
+	"net"
+)
+
+// IPFamily refers to a specific family if not empty, i.e. "4" or "6".
+type IPFamily string
+
+// Constants for valid IPFamilys:
+const (
+	IPFamilyUnknown IPFamily = ""
+
+	IPv4 IPFamily = "4"
+	IPv6 IPFamily = "6"
+)
+
+// IsDualStackIPs returns if a slice of ips is:
+// - all are valid ips
+// - at least one ip from each family (v4 or v6)
+func IsDualStackIPs(ips []net.IP) (bool, error) {
+	v4Found := false
+	v6Found := false
+	for _, ip := range ips {
+		if ip == nil {
+			return false, fmt.Errorf("ip %v is invalid", ip)
+		}
+
+		if v4Found && v6Found {
+			continue
+		}
+
+		if IsIPv6(ip) {
+			v6Found = true
+			continue
+		}
+
+		v4Found = true
```

Comment:
```diff
diff --git a/net/net.go b/net/net.go
index b7c08e2..b500f4c 100644
--- a/net/net.go
+++ b/net/net.go
@@ -22,6 +22,7 @@ import (
 	"math"
 	"math/big"
 	"net"
+	"net/netip"
 	"strconv"
 )
 
@@ -113,49 +114,61 @@ func IsDualStackCIDRStrings(cidrs []string) (bool, error) {
 	return IsDualStackCIDRs(parsedCIDRs)
 }
 
-// IsIPv6 returns if netIP is IPv6.
-func IsIPv6(netIP net.IP) bool {
-	return netIP != nil && netIP.To4() == nil
+// IsIPv6 tests whether ip is IPv6.  Not being IPv6 does not imply that it is
+// necessarily IPv4.
+func IsIPv6(ip net.IP) bool {
+	nip, ok := netip.AddrFromSlice(ip)
+	if !ok {
+		return false
+	}
+	nip = nip.Unmap()
+	return nip.Is6()
 }
 
-// IsIPv6String returns if ip is IPv6.
+// IsIPv6String tests whether ip is IPv6.  Not being IPv6 does not imply that
+// it is necessarily IPv4.
 func IsIPv6String(ip string) bool {
-	netIP := ParseIPSloppy(ip)
-	return IsIPv6(netIP)
+	return IsIPv6(ParseIPSloppy(ip))
 }
 
-// IsIPv6CIDRString returns if cidr is IPv6.
-// This assumes cidr is a valid CIDR.
+// IsIPv6CIDRString tests whether cidr is IPv6.  Not being IPv6 does not imply
+// that it is necessarily IPv4.
 func IsIPv6CIDRString(cidr string) bool {
 	ip, _, _ := ParseCIDRSloppy(cidr)
 	return IsIPv6(ip)
 }
 
-// IsIPv6CIDR returns if a cidr is ipv6
+// IsIPv6CIDR tests whether cidr is IPv6.  Not being IPv6 does not imply
+// that it is necessarily IPv4.
 func IsIPv6CIDR(cidr *net.IPNet) bool {
-	ip := cidr.IP
-	return IsIPv6(ip)
+	return IsIPv6(cidr.IP)
 }
 
-// IsIPv4 returns if netIP is IPv4.
-func IsIPv4(netIP net.IP) bool {
-	return netIP != nil && netIP.To4() != nil
+// IsIPv4 tests whether ip is IPv6.  Not being IPv6 does not imply that it is
+// necessarily IPv6.
+func IsIPv4(ip net.IP) bool {
+	nip, ok := netip.AddrFromSlice(ip)
+	if !ok {
+		return false
+	}
+	nip = nip.Unmap()
+	return nip.Is4()
 }
 
-// IsIPv4String returns if ip is IPv4.
+// IsIPv4String tests whether ip is IPv4.  Not being IPv4 does not imply that it is
+// necessarily IPv4.
 func IsIPv4String(ip string) bool {
-	netIP := ParseIPSloppy(ip)
-	return IsIPv4(netIP)
+	return IsIPv4(ParseIPSloppy(ip))
 }
 
-// IsIPv4CIDR returns if a cidr is ipv4
+// IsIPv4CIDR tests whether cidr is IPv4.  Not being IPv4 does not imply that it is
+// necessarily IPv4.
 func IsIPv4CIDR(cidr *net.IPNet) bool {
-	ip := cidr.IP
-	return IsIPv4(ip)
+	return IsIPv4(cidr.IP)
 }
 
-// IsIPv4CIDRString returns if cidr is IPv4.
-// This assumes cidr is a valid CIDR.
+// IsIPv4CIDRString tests whether cidr is IPv4.  Not being IPv4 does not imply
+// that it is necessarily IPv4.
 func IsIPv4CIDRString(cidr string) bool {
 	ip, _, _ := ParseCIDRSloppy(cidr)
 	return IsIPv4(ip)
```

Reply:
OK, yes, `IsIPv6(net.IP([]byte{}))` will return `true`. But I believe:
- `IsIPv4()` correctly handles all inputs
- `IsIPv6()` correctly handles all valid `net.IP`s, and `nil`
- `IsIPv4String()` and `IsIPv6String()` correctly handle all inputs (returning `false` for anything that can't be converted to a valid `net.IP`)

we can fix `IsIPv6()` to DTRT with invalid `net.IPs` too if you want. I don't think there's any need to drag in `netip`.

Reply:
Yeah, my little hack test was using `ip.To16()` rather than `!ip.To4()` which was poor fidelity.  Better testing shows I was wrong.

https://go.dev/play/p/K9H0pJx8eCg

The only one that is odd (and both impls get it wrong) is  "::FFFF:C0A8:1" and I am not sure how much we care?

I also wrote a benchmark for this, and it seems netip is BARELY faster:

```go
// You can edit this code!
// Click here and start typing.
package main

import (
	"net"
	"net/netip"
	"testing"
)

//go:noinline
func curIsIPv6(netIP net.IP) bool {
	return netIP != nil && netIP.To4() == nil
}

//go:noinline
func curIsIPv4(netIP net.IP) bool {
	return netIP != nil && netIP.To4() != nil
}

func addrFromIP(ip net.IP) netip.Addr {
	nip, ok := netip.AddrFromSlice(ip)
	if !ok {
		return netip.Addr{}
	}
	return nip.Unmap()
}

//go:noinline
func newIsIPv4(ip net.IP) bool {
	return addrFromIP(ip).Is4()
}

//go:noinline
func newIsIPv6(ip net.IP) bool {
	return addrFromIP(ip).Is6()
}

func do(b *testing.B, fam int) {
	cases := []struct {
		str string
		is4 bool
		is6 bool
	}{
		{"", false, false},
		{"0.0.0.0", true, false},
		{"127.0.0.1", true, false},
		{"::FFFF:C0A8:1", false, true},
		{"::1", false, true},
	}
	for _, tc := range cases {
		ip := net.ParseIP(tc.str)

		for i := 0; i < b.N; i++ {
			if fam == 4 {
				_ = curIsIPv4(ip)
				_ = curIsIPv6(ip)
			} else if fam == 6 {
				_ = newIsIPv4(ip)
				_ = newIsIPv6(ip)
			}
		}
	}
}

func BenchmarkCur(b *testing.B) {
	b.ReportAllocs()
	do(b, 4)
}

func BenchmarkNew(b *testing.B) {
	b.ReportAllocs()
	do(b, 6)
}
```

```
BenchmarkCur-6   	18189759	        57.13 ns/op	       0 B/op	       0 allocs/op
BenchmarkNew-6   	19459976	        53.26 ns/op	       0 B/op	       0 allocs/op
```